extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js

```diff
@@ -924,6 +975,61 @@ export class AmpStoryAutoAds extends AMP.BaseElement {
     return AD_STATE.INSERTED;
   }
 
+  /**
+   *
+   * @param {Element} adPageElement
+   */
+  maybeCreateAdChoices_(adPageElement) {
+    const {iframe} = this.lastCreatedAdImpl_;
+    const iframeDoc = getFrameDoc(iframe);
+    const href = getContentFromMetaTag(iframeDoc, Meta.ATTRIBUTION_URL);
+    const src = getContentFromMetaTag(iframeDoc, Meta.ATTRIBUTION_ICON);
+
+    // Ad Choices is optional, but need both to render.
+    if (!href && !src) {
+      return;
+    }
+
+    if (!href || !src) {
+      dev().warn('Both icon and URL must be supplied for Ad Choices.');
+      return;
+    }
+
+    const root = createElementWithAttributes(
+      this.doc_,
+      'div',
+      dict({
+        'role': 'button',
+      })
+    );
+
+    const adChoicesIcon = createElementWithAttributes(
+      this.doc_,
+      'img',
+      dict({
+        'class': 'i-amphtml-ad-choices-icon',
+        'src': src,
+      })
+    );
+
+    adChoicesIcon.addEventListener(
+      'click',
+      this.handleAdChoicesClick_.bind(this, href)
+    );
+
+    createShadowRootWithStyle(root, adChoicesIcon, adChoicesCSS);
```

Comment:
i might missed it in discussion. why shadow root is needed?

Reply:
Mostly CSS encapsulation. Before moving things to shadow we had a few bugs where pubs were breaking positioning, style etc.